src/expressions/if-expr.md

```diff
@@ -148,27 +103,71 @@ if let E::X(n) | E::Y(n) = v {
 }
 ```
 
-r[expr.if.let.lazy-bool]
-The expression cannot be a [lazy boolean operator expression][expr.bool-logic].
-Use of a lazy boolean operator is ambiguous with a planned feature change of the language (the implementation of if-let chains - see [eRFC 2947][_eRFCIfLetChain_]).
-When lazy boolean operator expression is desired, this can be achieved by using parenthesis as below:
+r[expr.if.chains]
+## Chains of conditions
+
+r[expr.if.chains.intro]
+Multiple condition operands can be separated with `&&`.
+
+r[expr.if.chains.order]
+Similar to a `&&` [_LazyBooleanOperatorExpression_], each operand is evaluated from left-to-right until an operand evaluates as `false` or a `let` match fails,
+in which case the subsequent operands are not evaluated.
+
+r[expr.if.chains.bindings]
+The bindings of each pattern are put into scope to be available for the next condition operand and the consequent block.
+
+The following is an example of chaining multiple expressions, mixing `let` bindings and boolean expressions, and with expressions able to reference pattern bindings from previous expressions:
 
-<!-- ignore: pseudo code -->
-```rust,ignore
-// Before...
-if let PAT = EXPR && EXPR { .. }
+```rust
+fn single() {
+    let outer_opt = Some(Some(1i32));
+
+    if let Some(inner_opt) = outer_opt
+        && let Some(number) = inner_opt
+        && number == 1
+    {
+        println!("Peek a boo");
+    }
+}
+```
 
-// After...
-if let PAT = ( EXPR && EXPR ) { .. }
+The above is equivalent to the following without using chains of conditions:
 
-// Before...
-if let PAT = EXPR || EXPR { .. }
+```rust
+fn nested() {
+    let outer_opt = Some(Some(1i32));
+
+    if let Some(inner_opt) = outer_opt {
+        if let Some(number) = inner_opt {
+            if number == 1 {
+                println!("Peek a boo");
+            }
+        }
+    }
+}
+```
+
+r[expr.if.chains.or]
+If any condition operand is a `let` pattern, then none of the condition operands can be a `||` [lazy boolean operator expression][_LazyBooleanOperatorExpression_] due to ambiguity and precedence with the `let` scrutinee.
+If a `||` expression is needed, then parentheses can be used. For example:
 
-// After...
-if let PAT = ( EXPR || EXPR ) { .. }
+```rust
+# let foo = Some(123);
+# let condition1 = true;
+# let condition2 = false;
+// Parentheses are required here.
+if let Some(x) = foo && (condition1 || condition2) { /*...*/ }
 ```
 
-[_eRFCIfLetChain_]: https://github.com/rust-lang/rfcs/blob/master/text/2497-if-let-chains.md#rollout-plan-and-transitioning-to-rust-2018
-[`match` expression]: match-expr.md
+r[expr.if.edition2024]
+> [!EDITION-2024]
+> Before the 2024 edition, let chains are not supported and only a single _IfCondition_ is allowed in an `if` expression.
```

Comment:
IIUC, this will need to be duplicate in loop-expr.md for `while` loops, too, correct?
(That is, until that restriction is removed.)

Reply:
when I wrote https://github.com/rust-lang/rust/pull/140204#issuecomment-2824605396 I thought so too, but there is this paragraph in the `while let patterns` section:

> Multiple condition operands can be separated with `&&`.
> These have the same semantics and restrictions as [`if` condition chains].

I'd say that this one implies that while let chains are under the same edition specific restriction.